RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc

```diff
@@ -202,35 +239,215 @@ void CAHitQuadrupletGenerator::hitQuadruplets(const TrackingRegion& region,
         hitQuadruplets(region, result, hitDoubletsPtr, g, es);
         theLayerCache.clear();
 }
-
-void CAHitQuadrupletGenerator::hitNtuplets(const IntermediateHitDoublets::RegionLayerSets& regionLayerPairs,
-                                              OrderedHitSeeds& result,
+void CAHitQuadrupletGenerator::hitNtuplets(const IntermediateHitDoublets& regionDoublets,
+                                              std::vector<OrderedHitSeeds, std::allocator<OrderedHitSeeds> >& result,
                                               const edm::EventSetup& es,
                                               const SeedingLayerSetsHits& layers) {
   CAGraph g;
 
+
+
   std::vector<const HitDoublets *> hitDoublets;
 
   auto layerPairEqual = [](const IntermediateHitDoublets::LayerPairHitDoublets& pair,
                            SeedingLayerSetsHits::LayerIndex inner,
                            SeedingLayerSetsHits::LayerIndex outer) {
     return pair.innerLayerIndex() == inner && pair.outerLayerIndex() == outer;
   };
-  fillGraph(layers, g, hitDoublets,
-            [&](const SeedingLayerSetsHits::SeedingLayer& inner,
-                const SeedingLayerSetsHits::SeedingLayer& outer,
-                std::vector<const HitDoublets *>& hitDoublets) {
-      using namespace std::placeholders;
-      auto found = std::find_if(regionLayerPairs.begin(), regionLayerPairs.end(),
-                                std::bind(layerPairEqual, _1, inner.index(), outer.index()));
-      if(found != regionLayerPairs.end()) {
-        hitDoublets.emplace_back(&(found->doublets()));
-        return true;
-      }
-      return false;
-    });
 
-  hitQuadruplets(regionLayerPairs.region(), result, hitDoublets, g, es);
+ int index =0;
+ for(const auto& regionLayerPairs: regionDoublets) {
+
+	 const TrackingRegion& region = regionLayerPairs.region();
+	 hitDoublets.clear();
+
+	  if (index == 0){   
+	  	createGraphStructure(layers, g);
+	  }
+	  else{  
+		clearGraphStructure(layers, g);
+	  }
+
+	  fillGraph(layers, g, hitDoublets,
+	            [&](const SeedingLayerSetsHits::SeedingLayer& inner,
+	                const SeedingLayerSetsHits::SeedingLayer& outer,
+	                std::vector<const HitDoublets *>& hitDoublets) {
+	      using namespace std::placeholders;
+	      auto found = std::find_if(regionLayerPairs.begin(), regionLayerPairs.end(),
+	                                std::bind(layerPairEqual, _1, inner.index(), outer.index()));
+	      if(found != regionLayerPairs.end()) {
+	        hitDoublets.emplace_back(&(found->doublets()));
+	        return true;
+	      }
+	      return false;
+	    });
+	edm::ESHandle<TrackerTopology> tTopoHand;
+	es.get<TrackerTopologyRcd>().get(tTopoHand);
+	const TrackerTopology *tTopo=tTopoHand.product();	
+
+
+	const int numberOfHitsInNtuplet = 4;
+	std::vector<CACell::CAntuplet> foundQuadruplets;
```

Comment:
Move this out of the loop, add `foundQuadruplets.clear()` for each iteration. (then the memory is re-used)